### DISCUSSION TITLE
USB support?

### DISCUSSION THREAD

**Discussion by T-vK** [pschatzmann/arduino-midi#9] [Q&A] [answered]
Thank you for this amazing library! Are there any plans on adding support for USB?
Many Arduino boards support acting as a USB MIDI controller: https://github.com/arduino-libraries/MIDIUSB
And for ESPs there's also a library for this: https://github.com/adafruit/Adafruit_TinyUSB_Arduino

Some ESPs like the ESP32-S3 even support host USB, meaning you could connect a standard USB midi keyboard to it and for Arduinos there are USB host shields to add support for this.

It would be great if support for those scenarios could be added at some point.

**Comment by pschatzmann** [comment]
Have a look at [my fork](https://github.com/pschatzmann/Adafruit_TinyUSB_Arduino/tree/Audio): It should work with any processor which has a standard integration of TinyUSB.

Unfortunately the ESP32 is a big mess and I am not sure if I will be able to find a solution there...

**Reply by pschatzmann** [reply, marked as answer]
Oops, I just realize that this is not for audio but for midi!

**Reply by T-vK** [reply]
Oh I see, that would explain why I couldn't find any commits about Midi 😅

**Comment by pschatzmann** [comment]
Did you try to use it with the TinyUSB library ? Adafruit_USBD_MIDI extends from Stream: So you can use it as input or output

**Reply by T-vK** [reply]
Oh that's good to know, thanks I wasn't aware. I will try it.
